### PR TITLE
feat: [M13] N4: Prefill via parallel scan in the Swift engine

### LINKED ISSUE #169: [M13] N4: Prefill via parallel scan in the Swift engine
Part of #163.

Carry the parallel‑scan prefill into the Swift engine so the prompt is ingested in one pass instead of token‑by‑token — the biggest single‑stream latency win.

## Scope
- Implement `prefill(tokens) -> (logits, State)` in Swift, mirroring the Python seam method (#165): run the SSD chunked scan + batched attention over the prompt, extract the recurrent `State` (Mamba conv window + final ssm_state `new_states[:, -1]` read at L‑1; attention K/V caches; MoE placeholder), then hand off to the mlx‑swift‑lm decode loop.

## Acceptance criteria
- Swift `prefill` + decode == Swift pure step‑by‑step decode, fp32 ~1e‑4.
- Swift `prefill` state/logits == the Python engine's `prefill` (#165) on the same prompt/weights, fp32 ~1e‑4.
- Measured prefill latency win vs the token‑by‑token baseline (feeds the bench, #N5).

## Dependencies
Blocked by #166 (N1 port) and #165 (Python prefill seam / parity oracle).

### PR DESCRIPTION
**Pull request by travisgalloway**
Closes #169

## Summary
- Adds `MonicaModel.prefill(tokens, lastOnly:)` — a one-shot SSD chunked-matmul scan over the whole prompt that hands `Generator` the exact recurrent `State` an L-step `step()` walk would have left: `SelectiveSSM.parallelWithState` (surfaces the carry-out the scan already computes and previously discarded), `convWindow`, and `forwardPrefill` on `MambaBlock`/`AttentionBlock`/`MoEBlock`.
- `Generator.generate` defaults to `usePrefill: true`; the old sequential per-token prefill stays reachable via `usePrefill: false` / `monica-generate --no-prefill` as the AC3 baseline and AC1 A/B rollback.
- `monica-parity` gains checks P1-P5: prefill logits vs the Python oracle and vs `forward`; `lastOnly` honesty; **state handoff, element-wise, both cross-language (vs a new `prefill.safetensors` oracle) and intra-Swift (vs `step`'s own recurrence)** — the load-bearing check, since a wrong carry-out state is silent in the prompt's own logits and only corrupts tokens generated afterward; prefill-then-decode vs pure step-by-step; exact greedy-id equality through the prefill path. The fp32 `rtol=1e-4`/`atol=1e-5` gate is unchanged.
- New `toy-short` fixture (B=1, L=2) exercises `convWindow`'s LEFT-PAD branch, which no existing fixture reaches (`d_conv=4`, so `L < d_conv-1`).
- Python exporter (`scripts/export_parity_fixture.py`) now refuses to write a fixture whose own `check_prefill_decode_parity` fails, and writes `prefill.safetensors` (`prefill_logits`/`prefill_last_logits`/`state.{i}.{slot}`) for every fixture. All 7 existing fixtures regenerated — only `meta.json` (new diagnostic fields) and the new `prefill.safetensors` changed; `weights.safetensors`/`reference.safetensors`/`generation.safetensors` are byte-identical.
- `monica-generate` gains `--bench-prefill` (AC3: sequential-vs-parallel prefill latency, asserts argmax agreement, wired into CI's `swift-engine` job as an informational/non-gating step) and `--no-prefill`.
- `swift/Package.swift` is untouched (`git diff main -- swift/Package.swift` is empty) — this is entirely a `swift/engine/` change.

## Testing
- [x] `.venv/bin/python -m pytest -q -rs` — 1572 passed, 13 skipped (pre-existing environment skips: no CUDA, no torch.compile toolchain, no prettier), 0 failed.
- [x] `.venv/bin/python -m pytest tests/test_mlx_parity.py tests/test_parity_fixture_export.py -q` — 14 passed, including the extended fixture-staleness check for `prefill.safetensors`.
- [x] `.venv/bin/python scripts/smoke_test.py --data <fresh toy split>` — PASSED (resume-exact, eval, and its own `[prefill]` check).
- [x] `cd swift && swift build && swift run monica-selfcheck` — OK, tokenizer package unaffected.
- [x] `cd swift/engine && swift build` — compiles clean (both `monica-generate` and `monica-parity`).
- [x] `swift run monica-generate --self-test` — OK, including new flag-parsing checks for `--no-prefill`/`--bench-prefill`.
- [ ] **CI-only** (this host has Command Line Tools only, no Xcode — `MONICA_ENGINE_CPU=1 swift run monica-parity` fails with "Failed to load the default metallib", confirmed pre-existing and unrelated to this change): the `swift model parity (macOS, mlx-swift)` CI job running `monica-parity`'s P1-P5 checks, `monica-generate`'s self-test/AC2/AC3 steps under `xcodebuild`, and the `--bench-prefill` informational step. AC3's measured latency numbers will be recorded in `docs/design/14-inference-engine.md` and as an issue comment once this job is green.

### CODE REVIEW

File: swift/engine/Sources/MonicaEngine/MonicaModel.swift [hunk 1]
```diff
@@ -97,6 +97,35 @@ public final class MonicaModel: Module {
...
+    ///
+    /// Fresh-session-only and no `seg_ids`, mirroring #165's seam contract
+    /// (`src/model/interface.py:81-95`): the attention RoPE positions are seeded from 0, so
+    /// a prompt may not be appended to a session that has already consumed tokens, and there
+    /// is no packing-aware carry-masking here. Walks `layers` directly (not through any
+    /// grad-checkpoint closure — that is a training concern; prefill is inference, matching
+    /// Python's `prefill` which also bypasses `_layer_fns`).
+    public func prefill(_ tokens: MLXArray, lastOnly: Bool = false) -> (MLXArray, [LayerState]) {
+        var h = cast(embedding(tokens), config.cd)
+        var state: [LayerState] = []
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply swift/engine/Sources/MonicaEngine/MonicaModel.swift:117]: ✅ Fixed in a5b5661d4c45b92434d48824c2da410c2ad36a09 — prefill now preconditions tokens are rank-2 with L > 0 before the lastOnly slice, failing loud instead of underflowing to (l-1) = -1 on an empty prompt.

File: swift/engine/Sources/monica-generate/main.swift [hunk 2]
```diff
@@ -242,14 +256,90 @@ func runSelfTest(tokenizerPath: String?) -> Never {
...
+/// `model.prefill`, timed over N iterations after a warm-up (lazy MLX graphs otherwise let
+/// the first call's compile/dispatch cost leak into the measurement). Asserts the two
+/// paths' argmax id AGREES — a correctness property that is not timing-sensitive — and
+/// prints per-path ms plus the speedup. Informational: CI wires this as a non-gating step
+/// (hosted-runner timing is noisy), so this function never calls `exit(1)` on timing alone.
+func runBenchPrefill(model: MonicaModel, promptLen: Int, iterations: Int) {
+    let vocab = model.config.vocabSize
+    let clampedLen = max(1, min(promptLen, vocab))
+    var rng = SystemRandomNumberGenerator()
+    let promptIds = (0..<clampedLen).map { _ in Int.random(in: 0..<vocab, using: &rng) }
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply swift/engine/Sources/monica-generate/main.swift:287]: ✅ Fixed in a5b5661d4c45b92434d48824c2da410c2ad36a09 — dropped the erroneous `min(promptLen, vocab)` clamp (prompt length is unrelated to vocab size and was silently truncating long prompts on small-vocab fixtures); also guarded `iterations` against 0 to avoid a divide-by-zero in the timing calc.

File: swift/engine/Sources/MonicaEngine/SelectiveSSM.swift [hunk 3]
```diff
@@ -159,7 +177,11 @@ public final class SelectiveSSM: Module {
 
         var y = split((ydiag + yoff).reshaped([bSize, lp, h, p]), indices: [l], axis: 1)[0]
         y = y + split(xh, indices: [l], axis: 1)[0] * d.reshaped([1, 1, h, 1])   // skip
-        return cast(y.reshaped([bSize, l, dInner]), cd)
+        // The carry-out is the OTHER half of the split that produced sEnter above: the
+        // entering state one chunk past the end, i.e. newStates[:, -1] — fp32, never cast
+        // to cd (state stays fp32, matching `recurrence`).
+        let carry = split(newStates, indices: [nc], axis: 1)[1].squeezed(axis: 1)   // (B,H,P,N)
+        return (cast(y.reshaped([bSize, l, dInner]), cd), carry)
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply swift/engine/Sources/MonicaEngine/SelectiveSSM.swift:187]: ✅ Fixed in a5b5661d4c45b92434d48824c2da410c2ad36a09 — split `newStates` once into `newStatesSplit` and reused both halves for `sEnter` and `carry` instead of calling split twice at the same index.

**Comment by travisgalloway** [comment]
Addressed review feedback:

- @copilot-pull-request-reviewer: all 3 flagged line comments fixed in a5b5661 — `prefill` now preconditions a non-empty rank-2 prompt before the `lastOnly` slice, the prefill bench no longer clamps prompt length to vocab size (and guards `iterations` against 0), and `SelectiveSSM.scan` splits `newStates` once and reuses both halves instead of splitting twice.
